marlowe-runtime/marlowe-contract/Main.hs

```diff
@@ -0,0 +1,213 @@
+      { loadSource = SomeConnectionSourceTraced inject
+          $ handshakeConnectionSourceTraced loadSource
+      , contractStore
+      , batchSize = unsafeIntToNat bufferSize
+      }
+
+    probeServer -< ProbeServerDependencies { port = fromIntegral httpPort, .. }
+
+runAppM :: EventBackend IO r RootSelector -> AppM r a -> IO a
+runAppM eventBackend = flip runReaderT (hoistEventBackend liftIO eventBackend) . unAppM
+
+newtype AppM r a = AppM
+  { unAppM :: ReaderT (EventBackend (AppM r) r RootSelector) IO a
+  } deriving newtype (Functor, Applicative, Monad, MonadIO, MonadUnliftIO, MonadFail, MonadThrow, MonadCatch, MonadMask)
+
+instance MonadWith (AppM r) where
+  type WithException (AppM r) = WithException IO
+  stateThreadingGeneralWith
+    :: forall a b releaseReturn
+     . GeneralAllocate (AppM r) (WithException IO) releaseReturn b a
+    -> (a -> AppM r b)
+    -> AppM r (b, releaseReturn)
+  stateThreadingGeneralWith (GeneralAllocate allocA) go = AppM . ReaderT $ \r -> do
+    let
+      allocA' :: (forall x. IO x -> IO x) -> IO (GeneralAllocated IO (WithException IO) releaseReturn b a)
+      allocA' restore = do
+        let
+          restore' :: forall x. AppM r x -> AppM r x
+          restore' mx = AppM . ReaderT $ restore . (runReaderT . unAppM) mx
+        GeneralAllocated a releaseA <- (runReaderT . unAppM) (allocA restore') r
+        let
+          releaseA' relTy = (runReaderT . unAppM) (releaseA relTy) r
+        pure $ GeneralAllocated a releaseA'
+    stateThreadingGeneralWith (GeneralAllocate allocA') (flip (runReaderT . unAppM) r . go)
+
+instance MonadEvent r RootSelector (AppM r) where
+  askBackend = askBackendReaderT AppM id
+  localBackend = localBackendReaderT AppM unAppM id
+
+data Options = Options
+  { host :: HostName
+  , port :: PortNumber
+  , bufferSize :: Int
+  , contractStoreDirectory :: FilePath
+  , contractStoreStagingDirectory :: FilePath
+  , httpPort :: PortNumber
+  }
+
+getOptions :: IO Options
+getOptions = do
+  ContractStoreOptions{..} <- defaultContractStoreOptions
+  execParser $ info
+    ( helper <*>
+      ( Options
+          <$> hostParser
+          <*> portParser
+          <*> bufferSizeParser
+          <*> contractStoreDirectoryParser contractStoreDirectory
+          <*> contractStoreStagingDirectoryParser contractStoreStagingDirectory
+          <*> httpPortParser
+      )
+    )
+    infoMod
+  where
+    hostParser = strOption $ mconcat
+      [ long "host"
+      , short 'h'
+      , value "127.0.0.1"
+      , metavar "HOST_NAME"
+      , help "The host name to run the server on."
+      , showDefault
+      ]
+
+    portParser = option auto $ mconcat
+      [ long "port"
+      , short 'p'
+      , value 3727
+      , metavar "PORT_NUMBER"
+      , help "The port number to run the marlowe load server on."
+      , showDefault
+      ]
+
+    bufferSizeParser = option readOption $ mconcat
+      [ long "buffer-size"
+      , short 'b'
+      , value 4098
```

Comment:
Seems too high. Are exceptions caught in a way the forces a flush before exit?

Reply:
1024 could be a better default. The staging area is wrapped in a bracket that discards the entire temp staging directory on exception. So there should be no resource leak (though nothing will protect from a SIGKILL or a loss of power of course).